src/events/broker.py

```diff
@@ -482,8 +495,102 @@ def _format_storages(self) -> None:
         self.workload.format_storages(
             uuid=self.charm.state.peer_cluster.cluster_uuid,
             internal_user_credentials=credentials,
+            initial_controllers=f"{self.charm.state.peer_cluster.bootstrap_unit_id}@{self.charm.state.peer_cluster.bootstrap_controller}:{self.charm.state.peer_cluster.bootstrap_replica_id}",
         )
 
+    def _leader_elected(self, event: LeaderElectedEvent) -> None:
+        if self.charm.state.runs_controller and self.charm.state.cluster.bootstrap_controller:
+
+            updated_bootstrap_data = {
+                "bootstrap-controller": self.charm.state.bootstrap_controller,
+                "bootstrap-unit-id": str(self.charm.state.kraft_unit_id),
+                "bootstrap-replica-id": self.charm.state.unit_broker.directory_id,
+            }
+            self.charm.state.cluster.update(updated_bootstrap_data)
+
+            if self.charm.state.peer_cluster_orchestrator:
+                self.charm.state.peer_cluster_orchestrator.update(updated_bootstrap_data)
+
+            # A rolling restart is required to apply the new leader config
+            self.charm.on[f"{self.charm.restart.name}"].acquire_lock.emit()
```

Comment:
**todo:** We should remove this event handler. Let's just always attempt to update these relation-data items during normal `update-status`/`config-changed`, and let their reconciliation check for diffs in the config and restart normally.
Generally I'd like to avoid force-restarting wherever possible, and let the reconciliation do it automatically. Ends up being a bit more resilient, in case we miss a particular event somewhere in defer hell.

Reply:
Sorry, I think I need a little clarification here, do you mean I should remove the `leader_elected` hook altogether, or avoid emitting restart?

Reply:
Both, I think.
I think we should run this block of code every `config-changed` event:
```python
        if self.charm.state.runs_controller and self.charm.state.cluster.bootstrap_controller:

            updated_bootstrap_data = {
                "bootstrap-controller": self.charm.state.bootstrap_controller,
                "bootstrap-unit-id": str(self.charm.state.kraft_unit_id),
                "bootstrap-replica-id": self.charm.state.unit_broker.directory_id,
            }
            self.charm.state.cluster.update(updated_bootstrap_data)

            if self.charm.state.peer_cluster_orchestrator:
                self.charm.state.peer_cluster_orchestrator.update(updated_bootstrap_data)
```

Because there's no harm in doing it every time, right?

And then, when we check for diffs in the `*.properties/jaas` files, we will auto-emit the restart anyway. If you like we can chat about it and I can try to explain myself better.

Reply:
Replaced the redundant restart with a config changed and the leader_elected hook is now handled in controller event handler. Thanks!